test/provisioner/provision_ibmcloud.go

```diff
@@ -5,33 +5,455 @@ package provisioner
+	}
+
+	sgoptions := &vpcv1.GetVPCDefaultSecurityGroupOptions{}
+	sgoptions.SetID(IBMCloudProps.VpcID)
+	defaultSG, _, err := IBMCloudProps.VPC.GetVPCDefaultSecurityGroup(sgoptions)
+	if err != nil {
+		return err
+	}
+
+	IBMCloudProps.SecurityGroupID = *defaultSG.ID
+	ibmcloudTrace(fmt.Sprintf("Got VPC default SecurityGroupID %s.\n", IBMCloudProps.SecurityGroupID))
+
+	return nil
+}
+
+func deleteVPC() error {
+	if !IBMCloudProps.IsProvNewVPC {
+		fmt.Printf("Do not delete because using existing VPC: %s\n", IBMCloudProps.VpcID)
+		return nil
+	}
+
+	deleteVpcOptions := &vpcv1.DeleteVPCOptions{}
+	deleteVpcOptions.SetID(IBMCloudProps.VpcID)
+	ibmcloudTrace(fmt.Sprintf("Deleting VPC with ID %s.\n", IBMCloudProps.VpcID))
+	_, err := IBMCloudProps.VPC.DeleteVPC(deleteVpcOptions)
+
+	if err != nil {
+		return err
+	}
+	ibmcloudTrace(fmt.Sprintf("Deleted VPC with ID %s.\n", IBMCloudProps.VpcID))
+	return nil
+}
+
+func createSubnet() error {
+	if !IBMCloudProps.IsProvNewSubnet {
+		fmt.Printf("Using existing Subnet: %s\n", IBMCloudProps.SubnetID)
+		return nil
+	}
+
+	cidrBlock := getCidrBlock(IBMCloudProps.Region, IBMCloudProps.Zone)
+	if cidrBlock == "" {
+		return errors.New("Can not calculate cidrBlock from Region and Zone.")
+	}
+	options := &vpcv1.CreateSubnetOptions{}
+	options.SetSubnetPrototype(&vpcv1.SubnetPrototype{
+		ResourceGroup: &vpcv1.ResourceGroupIdentity{
+			ID: &IBMCloudProps.ResourceGroupID,
+		},
+		Ipv4CIDRBlock: &cidrBlock,
+		Name:          &[]string{IBMCloudProps.SubnetName}[0],
+		VPC: &vpcv1.VPCIdentity{
+			ID: &IBMCloudProps.VpcID,
+		},
+		Zone: &vpcv1.ZoneIdentity{
+			Name: &IBMCloudProps.Zone,
+		},
+	})
+	ibmcloudTrace(fmt.Sprintf("Creating subnet %s in VPC %s in Zone %s.\n", IBMCloudProps.SubnetName, IBMCloudProps.VpcID, IBMCloudProps.Zone))
+	subnet, _, err := IBMCloudProps.VPC.CreateSubnet(options)
+	if err != nil {
+		return err
+	}
+	IBMCloudProps.SubnetID = *subnet.ID
+	ibmcloudTrace(fmt.Sprintf("Created subnet with ID %s.\n", IBMCloudProps.SubnetID))
+
+	if len(IBMCloudProps.SubnetID) <= 0 {
+		return errors.New("SubnetID is empty, unknown error happened when create Subnet.")
+	}
+
+	return nil
+}
+
+func deleteSubnet() error {
+	if !IBMCloudProps.IsProvNewSubnet {
+		fmt.Printf("Do not delete because using existing Subnet: %s\n", IBMCloudProps.SubnetID)
+		return nil
+	}
+
+	options := &vpcv1.DeleteSubnetOptions{}
+	options.SetID(IBMCloudProps.SubnetID)
+	ibmcloudTrace(fmt.Sprintf("Deleting subnet with ID %s.\n", IBMCloudProps.SubnetID))
+	_, err := IBMCloudProps.VPC.DeleteSubnet(options)
+
+	if err != nil {
+		return err
+	}
+	ibmcloudTrace(fmt.Sprintf("Deleted subnet with ID %s.\n", IBMCloudProps.SubnetID))
+	return nil
+}
+
+func createVpcImpl() error {
+	err := createVPC()
+	if err != nil {
+		return err
+	}
+	ibmcloudTrace("waiting vpc ready before create subnet")
+	// wait vpc ready before create subnet
+	time.Sleep(60 * time.Second)
```

Comment:
In future we should make this check for the VPC to be available rather than just sleeping 60 secs.

Reply:
Sleep 60 seconds is good enough here, actually, if we don't use "manual" for AddressPrefixManagement when create the VPC, I think we don't even need the sleep. AddressPrefixManagement "manual" brings more flexibility but lost some functionality. So, we use the default AddressPrefixManagement with fixed cidr described here https://cloud.ibm.com/docs/vpc?topic=vpc-configuring-address-prefixes, I think it's good enough for e2e test.

Reply:
Yeah, my feedback was based on the fact I could see that it went to available almost immediately, so I figured it we had a smarter polling approach it would save the delay in the test run.